### DISCUSSION TITLE
How do I subscribe to error/timeout events on realtime channels?

### DISCUSSION THREAD

**Discussion by jonascript** [supabase/supabase#10293] [Questions]
I'm trying to add logic to my application to handle cases where the realtime subscription fails or errors unexpectedly. We actually had this happen about a day after deploying. 

Looking through the supabase realtime docs, I couldn't find a mention of error handling code.

Does supabase realtime, automatically recover from timeout and subscription errors?  

```
supabase
      .channel(`public:my_table`)
      .on(
        'postgres_changes',
        {
          event: 'UPDATE',
          schema: 'public',
          table: 'my_table',
        },
        (payload) => { 
             console.log('EVENT', payload);
        }
      )
      .on(
        'postgres_changes',
        {
          event: 'INSERT',
          schema: 'public',
          table: 'my_table',
        },
       (payload) => { 
             console.log('EVENT', payload);
        }
      )
      .subscribe();
```

Looking at this example, it appears as if I should be able to add "receive" methods to listen to these events but when trying on the main supabase js client, the method doesn't exist, which indicates that this is abstracted away. 

https://github.com/supabase/realtime/blob/954dfee803dfed43f53483e8fe84218bc87daeab/examples/node-js/src/server.js

I can add the "receive" method to the "joinPush" object that subscribe returns but this seems hacky.
```
  .subscribe()
      .joinPush.receive('ok', (event) => {
        console.log('OK', event)
      })
```

How can I add listeners to channel subscription errors so that we can have visibility into any unexpected errors?

**Comment by GaryAustin1** [comment]
Although realtime will attempt to recover from connection errors, you would lose any data change during those periods.
`.subscribe(status=>{console.log(status)})` is the main way to detect status of the connection. 

The readme here https://github.com/supabase/realtime-js  goes thru some of this.

**Reply by jonascript** [reply]
Thank you for the info @GaryAustin1. 

So if I understand correctly, this will fire asynchronously at any point the status changes, not just when attempting the initial subscribe.

If that is the case, do you have any recommendations on how implement recovery logic? 

channel.subscribe((status, err) => {
  if (status === 'SUBSCRIBED') {
    console.log('Connected!')
  }

  if (status === 'CHANNEL_ERROR') {
    console.log(`There was an error subscribing to channel: ${err.message}`)
  }

  if (status === 'TIMED_OUT') {
    console.log('Realtime server did not respond in time.')
  }

  if (status === 'CLOSED') {
    console.log('Realtime channel was unexpectedly closed.')
  }
})

**Reply by GaryAustin1** [reply]
I always load up my needed info from the monitored table with a select after I get subscribed event.
This discussion is a bit dated but it goes thru some of the things to think about: [How to obtain reliable realtime updates in the real world](https://github.com/supabase/supabase/discussions/5641)

**Reply by jonascript** [reply]
Thank you for sharing this @GaryAustin1. This is helpful though in my case I'm not dealing with tab visibility but this is good to know. 

In my case, I'm encountering an issue where intermittently there is a "CHANNEL_ERROR" subscribing to postgres changes on a node server. 

This has happened more than once. Do you have any ideas how to automatically recover if possible?

Would it just be a retry loop to subscribe again at an interval?

**Reply by GaryAustin1** [reply]
Normally realtime will do a retry with exponential backoff, (might depend on the type of error ).
But you would lose any changes during the reconnect periods.  If it is reconnecting successfully for you, you would reload your initial info when you get subscribed again as you may have lost stuff.  That is if you care about missing data when the subscriptions is down.

If it is not reconnecting on it's own, then a similar retry loop that get's slower each time might be the way to go.

It might be good for you to try and have an understanding on why you are losing connection, if it is not network issues.  Also if you are using RLS user tokens, you will have issues if you don't refresh them before they expire.
